llama-index-integrations/llms/llama-index-llms-zhipuai/llama_index/llms/zhipuai/base.py

```diff
@@ -0,0 +1,390 @@
+
+    return token_limit
+
+
+def is_function_calling_model(model: str) -> bool:
+    return "4v" not in model
+
+
+def force_single_tool_call(response: ChatResponse) -> None:
+    tool_calls = response.message.additional_kwargs.get("tool_calls", [])
+    if len(tool_calls) > 1:
+        response.message.additional_kwargs["tool_calls"] = [tool_calls[0]]
+
+
+def async_llm_generate(item):
+    try:
+        return next(item)
+    except StopIteration:
+        return None
+
+
+class ZhipuAI(FunctionCallingLLM):
+    """ZhipuAI LLM.
+
+    Visit https://open.bigmodel.cn to get more information about ZhipuAI.
+
+    Examples:
+        `pip install llama-index-llms-zhipuai`
+
+        ```python
+        from llama_index.llms.zhipuai import zhipuai
+
+        llm = ZhipuAI(model="glm-4", api_key="YOUR API KEY")
+
+        response = llm.complete("who are you?")
+        print(response)
+        ```
+    """
+
+    model: str = Field(description="The ZhipuAI model to use.")
+    api_key: Optional[str] = Field(
+        default=None,
+        description="The API key to use for the ZhipuAI API.",
+    )
+    temperature: float = Field(
+        default=0.95,
+        description="The temperature to use for sampling.",
+        ge=0.0,
+        le=1.0,
+    )
+    max_tokens: int = Field(
+        default=1024,
+        description="The maximum number of tokens for model output.",
+        gt=0,
+        le=4096,
+    )
+    timeout: float = Field(
+        default=DEFAULT_REQUEST_TIMEOUT,
+        description="The timeout for making http request to ZhipuAI API server",
+    )
+    additional_kwargs: Dict[str, Any] = Field(
+        default_factory=dict, description="Additional kwargs for the ZhipuAI API."
+    )
+    _client: Optional[ZhipuAIClient] = PrivateAttr()
+
+    def __init__(
+        self,
+        model: str,
+        api_key: str,
+        temperature: float = 0.95,
+        max_tokens: int = 1024,
+        timeout: float = DEFAULT_REQUEST_TIMEOUT,
+        additional_kwargs: Optional[Dict[str, Any]] = None,
+        **kwargs: Any,
+    ) -> None:
+        additional_kwargs = additional_kwargs or {}
+        super().__init__(
+            model=model,
+            temperature=temperature,
+            max_tokens=max_tokens,
+            timeout=timeout,
+            additional_kwargs=additional_kwargs,
+            **kwargs,
+        )
+
+        self._client = ZhipuAIClient(api_key=api_key)
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "ZhipuAI_LLM"
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        """LLM metadata."""
+        return LLMMetadata(
+            context_window=glm_model_to_context_size(self.model),
+            num_output=DEFAULT_NUM_OUTPUTS,
+            model_name=self.model,
+            is_chat_model=True,
+            is_function_calling_model=is_function_calling_model(self.model),
+        )
+
+    @property
+    def _model_kwargs(self) -> Dict[str, Any]:
```

Comment:
if property, then I think you likely want to make this non "private" attribute, and thus remove the "_"?